File: three.c
#include <stdio.h>
/* Three Card Game Paradox
 * by Conor Gilmer
 */

/* globals */
int n = 0; /* number of Runs */

/* return the card corresponding to the number */
char card(int x) 
{
	char cards[] = {'A','B','C'};
	return(cards[x-1]);
} 

/* Three card game iteration */ 
int run () {

	int winning, chosen, remov;
	n++;
	printf("Run %d \t", n);
	winning = rand()%3 + 1;
	chosen = rand()%3 + 1;
	
	do {
		remov = rand()%3 + 1;
/*	printf("Removing no %d, (%d, %d) \n", remov, winning, chosen);*/
	} while ((chosen == remov) || (winning == remov));

	printf("Winning no %c \t", card(winning));
	printf("Chosen no %c \t", card(chosen));
	printf("Removing no %c \t", card(remov));

	/* sticking so have you won*/
	if (chosen == winning) {
		printf("Sticking Wins\n");
		return 0;
	} else {
		printf("Twisting Wins\n");
		return 1;
	}
}

int main()
{
	int twist = 0;
	int stick = 0;
	int games = 112; /* number of runs */
	int m;
	float twistPC = 0.0;
	float stickPC = 0.0;

	printf("Three Card Game Paradox\n");
	/* play the game games times */
	for (m = 1; m < games+1; m++)
	{
		if (run() ==0) 
			stick++; 
		else 
			twist++;
	}

	/* get the percentages of wins for switching and twisting */
	twistPC = (((float)twist/games) * 100.0);
	stickPC = (((float)stick/games) * 100.0);

	/* output results */
	printf("Results:\n");
	printf("Sticking wins: %d times (%f %)\n", stick, stickPC);
	printf("Twisting wins: %d times (%f %)\n", twist, twistPC);
	/* verdict */
	if (stick > twist)
		printf("It is better to stick with your original choice.\n");
	else
		printf("It is better to twist.\n");
	printf("The End\n");

	return 0;
}

File: three.pas
program three;

var 
winning, chosen, remove:integer;
stick, twist, games, num:cardinal;
stickR, twistR, gamesR:real;
stickpc, twistpc:real;
letter:char;

(* return a random card a number between 1 and 3 *)
function randomcard(cards:integer) : integer;
var
rvar:integer;
begin
	randomize;
	rvar := random(cards) + 1;	
(*	writeln('random = ', rvar);*)
	randomcard := rvar;
end;

(* play a game of selecting from three cards *)
procedure play;
begin
	writeln('in play ', num);

	winning := randomcard(3);
	chosen := randomcard(3);
	
	repeat
  	remove := randomcard(3);
	until (remove <> winning) and (remove <> chosen);

	writeln('winning = ', winning, ' | chosen = ', chosen, ' | remove = ', remove );

	if winning = chosen then
	begin
		stick := stick + 1;
	end
	else
		twist := twist + 1;
end;

(* Main Program Three Card Paradox *)
begin

	games := 100;

	writeln('Three Card Paradox');
	for num :=1 to games do
		play;
	writeln('Results:');
	stickR := stick;
	twistR := twist;
	gamesR := games;
	stickpc := (stickR/gamesR) *100.0;
	twistpc := (twistR/gamesR) *100.0;
	writeln('Game played ', games, ' times');
	writeln('Sticking wins ', stick, ' times (', stickpc:2:2, '%)');
	writeln('Changing wins ', twist, ' times (', twistpc:2:2, '%)');
	writeln('Verdict:');
	if stick < twist then
	begin
		writeln('Twisting wins more!');
	end
	else
		writeln('Sticking wins more!');
	writeLn('The End.');
end.
